### PR TITLE
Immediate Mode Line/Gizmo Drawing

### LINKED ISSUE #5619: Improved rendering (And documentation) for 2d and 3d Lines
## The Problem:
Currently, it is difficult to utilize basic lines in your bevy project. The best 2 options are to use bevy polyline, or to import the shader assets from bevy main into your project, neither of which are very obvious or clear to beginners, and it is not very clean or straightforward to have to either import new dependencies or shaders into your project. 

Additionally, they only work with 3d, being incompatible with 2d cameras. They also are incredibly difficult and unintuitive to work with, (With Bevy main's line shader) it is a challenge to even be able to move the lines during run time, and it has a lot of overhead, which makes it extremely difficult to even get to use without copying and pasting from this one example.

## Solution:
It would be immensely helpful to have functions built-in that allow spawning bundles for lines in both 3d and 2d, which work 'out of the box' without needing to import/copy+paste a bunch of things, with documentation to make it more intuitive to use, implemented similarly to the `shapes` geometry bundles, which you can easily spawn with a `spawn_bundle` (The attributes all labeled), and then very simply edit on the fly with just a simple mutable query, as opposed to the `lines` right now, which... I'm not quite sure how it is intended to be edited??.. uh.. or the bevy polylines which require all sorts of handles and assets, which you get_mut another handle then unwrap and it's a whole difficult process.

Thanks in advance!

### PR DESCRIPTION
**Pull request by devil-ira**
# Objective
Add a convenient immediate mode drawing API for visual debugging.

Fixes #5619
Alternative to #1625
Partial alternative to #5734

Based off https://github.com/Toqozz/bevy_debug_lines with some changes:
 * Simultaneous support for 2D and 3D.
 * Methods for basic shapes; circles, spheres, rectangles, boxes, etc.
 * 2D methods.
 * Removed durations. Seemed niche, and can be handled by users.

<details>
<summary>Performance</summary>

Stress tested using Bevy's recommended optimization settings for the dev profile with the 
following command.
```bash
cargo run --example many_debug_lines \
    --config "profile.dev.package.\"*\".opt-level=3" \
    --config "profile.dev.opt-level=1"
```
I dipped to 65-70 FPS at 300,000 lines
CPU: 3700x
RAM Speed: 3200 Mhz
GPU: 2070 super - probably not very relevant, mostly cpu/memory bound

</details>

<details>
<summary>Fancy bloom screenshot</summary>

Screenshot_20230207_155033

</details>

## Changelog
 * Added `GizmoPlugin`
 * Added `Gizmos` system parameter for drawing lines and wireshapes.

### TODO
- [ ] Update changelog
- [x] Update performance numbers
- [x] Add credit to PR description

### Future work
- Cache rendering primitives instead of constructing them out of line segments each frame.
- Support for drawing solid meshes
- Interactions. (See [bevy_mod_gizmos](https://github.com/LiamGallagher737/bevy_mod_gizmos))
- Fancier line drawing. (See [bevy_polyline](https://github.com/ForesightMiningSoftwareCorporation/bevy_polyline))
- Support for `RenderLayers`
- Display gizmos for a certain duration. Currently everything displays for one frame (ie. immediate mode)
- Changing settings per drawn item like drawing on top or drawing to different `RenderLayers`  

Co-Authored By: @lassade <felipe.jorge.pereira@gmail.com>
Co-Authored By: @The5-1 <agaku@hotmail.de> 
Co-Authored By: @Toqozz <toqoz@hotmail.com>
Co-Authored By: @nicopap <nico@nicopap.ch>

### CODE REVIEW

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (james7132) [inline crates/bevy_debug_draw/src/debug_draw.rs:9]: Docs should openly specify how long the debug lines last on screen. One frame? Multiple?

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (james7132) [inline crates/bevy_debug_draw/src/debug_draw.rs:13]: Almost tempted to ask this to be put into a `Mutex` to allow multiple systems to draw at the same time. Otherwise, a `ResMut` on this will force all debug-enabled systems to serialize.

File: crates/bevy_internal/Cargo.toml
Comment (james7132) [inline crates/bevy_internal/Cargo.toml:61]: What does this `?` do?

File: crates/bevy_debug_draw/src/lib.rs
Comment (james7132) [inline crates/bevy_debug_draw/src/lib.rs:81]: Could this be merged into `DebugDraw` itself?

File: crates/bevy_debug_draw/src/lib.rs
Comment (james7132) [inline crates/bevy_debug_draw/src/lib.rs:128]: This seems to be spawned every game tick and never destroyed. Do we really need an entity for this?

**Review by james7132 — COMMENTED** [review]
Bikeshedding for a better long term name: `bevy_gizmos`? It'd fit better into the context of a user-facing editor.

**Comment by james7132** [comment]
Credit should be given to @lassade, @The5-1, and @Toqozz for their prior work in this space.

Could you add `Co-Authored By: <EMAIL>` to the PR description? Bors will merge it in with that in the final commit to main.

Separately, @lassade's existing implementation for `bevy_gizmos` seems to also have support for filling in the internal space with a flat color. We can add it in a follow-up PR or leave it be.

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (james7132) [inline crates/bevy_debug_draw/src/debug_draw.rs:90]: Reserving the space in positions and colors here would avoid potential reallocations.

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (james7132) [inline crates/bevy_debug_draw/src/debug_draw.rs:193]: Does this need to be `pub`? Can it just be `pub(crate)`?

Alternatively, could we embed the Mesh here so it's edited in place instead of batch copied? This would avoid the extra copy.

**Review by james7132 — CHANGES_REQUESTED** [review]

**Comment by The5-1** [comment]
Oh I totally missed this here, good stuff 👌

Was still struggling to get my old attempt ported to 0.8 and then 0.9 hit ;p
Gotta take a closer look at this tomorrow to see how its done!

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (nicopap) [reply crates/bevy_debug_draw/src/debug_draw.rs:13]: Note that <https://github.com/nicopap/bevy-debug-text-overlay/> does just this. Though it uses a mpsc sync_channel, and has a system that reads the receiver once per frame. 

Another advantage of such a method is that you don't need to specify the `DebugLines` resource on the systems that use it

**Comment by nicopap** [comment]
Note that I also heavily contributed to [Toqozz/bevy_debug_lines](https://github.com/Toqozz/bevy_debug_lines). I did the port to the new renderer for 0.6, it was basically an entire rewrite.

File: crates/bevy_internal/Cargo.toml
Comment (devil-ira) [reply crates/bevy_internal/Cargo.toml:61]: Without the `?` enabling the `bevy_ci_testing` feature will unconditionally enable the optional `bevy_render` crate.
With it, `bevy_render/ci_limits` will only be enabled when `bevy_render` was already enabled.
It's unrelated to the rest of this PR.

File: crates/bevy_debug_draw/src/lib.rs
Comment (The5-1) [reply crates/bevy_debug_draw/src/lib.rs:128]: In my approach there is a mesh entity spawned during plugin setup and then the mesh data on it is updated every frame.
Could possibly be a solution for this.
(I currently would not know how to render without an mesh entity at all. So if there is a way that may be interesting too.)

Comment (nicopap) [reply crates/bevy_debug_draw/src/lib.rs:128]: Actually, this doesn't spawn one mesh per frame, it only does generate the mesh if there isn't one already (see line 112), otherwise it just updates the existing one.

File: crates/bevy_debug_draw/src/lib.rs
Comment (nicopap) [inline crates/bevy_debug_draw/src/lib.rs:134]: I think this is still missing a `NoFrustrumCulling` component, otherwise the debug lines will only be visible if the origin is in the field of view (due to #4294)

File: crates/bevy_internal/Cargo.toml
Comment (james7132) [reply crates/bevy_internal/Cargo.toml:61]: TIL. Great to know.

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (nicopap) [inline crates/bevy_debug_draw/src/debug_draw.rs:32]: I did implement a collection of outlines here if you want a reference. Though unlike this, which uses `PrimitiveTopology::LineList`, it assumes a `PrimitiveTopology::LineStrip`, which divide by two the number of required indices if your lines are continuous. <https://github.com/nicopap/bevy-cool-shapes/blob/d84c8efcb4bbccd188a1ca8752345b44a7812a7c/bevy_cool_shapes_render/src/outline.rs>

Comment (devil-ira) [reply crates/bevy_debug_draw/src/debug_draw.rs:32]: Totally forgot that it's possible to use NaNs to split a LineStrip!

File: crates/bevy_debug_draw/src/pipeline_2d.rs
Comment (devil-ira) [reply crates/bevy_debug_draw/src/pipeline_2d.rs:25]: The `sort_key` on `Transparent2d` needs to be set. Doesn't seem possible via `Material2d`

File: crates/bevy_debug_draw/src/lib.rs
Comment (devil-ira) [reply crates/bevy_debug_draw/src/lib.rs:128]: This code turned into a mess when I added the `config.enabled` branches.
A `match` is indeed much clearer :)

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (devil-ira) [reply crates/bevy_debug_draw/src/debug_draw.rs:193]: It doesn't seem like anything is getting copied?

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (devil-ira) [reply crates/bevy_debug_draw/src/debug_draw.rs]: I'll be testing both approaches to see how the performance and ergonomics compare :)

File: crates/bevy_debug_draw/src/lib.rs
Comment (devil-ira) [reply crates/bevy_debug_draw/src/lib.rs:84]: This was split because it's awkward to have private fields on a config struct like this. It also gets sent to the render world.

File: crates/bevy_debug_draw/src/lib.rs
Comment (devil-ira) [reply crates/bevy_debug_draw/src/lib.rs]: I've was thinking about that as well.
It's probably possible to spawn the entity directly in the render world each frame.

**Comment by alice-i-cecile** [comment]
Once #5303 by @IceSentry is merged, should we reuse the wireframe material here?

**Comment by devil-ira** [comment]
> Once https://github.com/bevyengine/bevy/pull/5303 by @IceSentry is merged, should we reuse the wireframe material here?

This PR needs colors to be per vertex with an option to disable depth testing + a 2d version, so that's not an option, and the `Material(2d)` traits and `AsBindGroup` are currently not flexible enough to support this.

**Comment by devil-ira** [comment]
Using a channel/mutex ended up rather costly for performance. I switched to the method `Commands` uses. I temporarily yoinked the code from #6817 for the draft :)

I'd like to give some more performance numbers but the systems in `many_debug_lines.rs` don't appear to be running in parallel.

Still some things to do, like renaming the crate, so I'd hold off on doing another review pass :)

File: crates/bevy_debug_draw/src/debug_draw.rs
Comment (devil-ira) [reply crates/bevy_debug_draw/src/debug_draw.rs]: An immediate mode drawing API is frame by frame, but I'll do a docs pass and write it down in plain language :)

**Comment by james7132** [comment]
@devil-ira #5619 seems to be addressed by this. Can you add `Fixes #5619` to the PR description?

**Comment by nicopap** [comment]
I think that the API as a system param is lacking and doesn't fulfill the goal of creating an easy API for debugging. An API to draw gizmos for editing is different than an API to draw lines for understanding what is going on in your code. Of course, it can share methods, but having different goals they should be useable differently.

Here is an example to illustrate what I mean. I have two functions and a system

```rust
struct Segment {
    move_left: bool,
    value: Vec2,
}
#[derive(Component)]
struct Segments(Vec<Vec<Segment>>);

fn move_segment(segment: &mut Segment) {
    if segment.move_left {
        segment.value += Vec2::X;
    }
}
fn move_list(segments: &mut [Segment]) {
    for segment in segments {
        move_segment(segment);
    }
}

fn system(mut segments: Query<&mut Segment>) {
    for segment_list in &mut segments {
        move_list(&mut segment_list.0);
    }
}
```

Now I want to draw a line **but only when the segment moved**. Well, with your API, I need to modify **all the parameter lists** to add this line.

```rust
struct Segment {
    move_left: bool,
    value: Vec2,
}
#[derive(Component)]
struct Segments(Vec<Vec<Segment>>);

fn move_segment(segment: &mut Segment, draw_gizmo: &mut DrawGizmo) {
    if segment.move_left {
        segment.value += Vec2::X;
        draw_gizmo.add_line(segment.value, segment.value + Vec2::Y);
    }
}

fn move_list(segments: &mut [Segment], draw_gizmo: &mut DrawGizmo) {
    for segment in segments {
        move_segment(segment, draw_gizmo);
    }
}

fn system(mut segments: Query<&mut Segment>, mut draw_gizmo: DrawGizmo) {
    for segment_list in &mut segments {
        move_list(&mut segment_list.0, &mut *draw_gizmo);
    }
}
```

For one, this is extremely annoying. Secondly, a common thing to do is to push the debug drawing behind a compile flag, how would I do it with your API? This is so unpleasant that it defeats the purpose of an immediate-mode drawing API.

Ideally the API looks like this:

```rust
struct Segment {
    move_left: bool,
    value: Vec2,
}
#[derive(Component)]
struct Segments(Vec<Vec<Segment>>);

fn move_segment(segment: &mut Segment) {
    if segment.move_left {
        segment.value += Vec2::X;
        DrawGizmo::add_line(segment.value, segment.value + Vec2::Y);
        // OR
        draw_line!(segment.value, segment.value + Vec2::Y);
    }
}
fn move_list(segments: &mut [Segment]) {
    for segment in segments {
        move_segment(segment);
    }
}

fn system(mut segments: Query<&mut Segment>) {
    for segment_list in &mut segments {
        move_list(&mut segment_list.0);
    }
}
```

(also credits to chatGPT for drafting the two first code samples)

**Comment by cart** [comment]
The current SystemParam approach is definitely the least controversial. It builds on existing patterns, it makes commands easily discoverable / autocomplete-able. and it doesn't introduce the weirdness that globals introduce (statefulness gets more challenging, it can break multi-app scenarios such as testing, harder for people to understand control flow, etc).

I think an argument could be made for a set of `println!` style apis, but:
1. That isn't implemented yet
2. We'd need to sort out the implications of doing it that way, work out the scenarios where it might break, and then weigh that against the value we get by doing it. I suspect we'd come out on top here, but thats "additional work".
3. It seems like they could exist in addition to the more standard approach used in this PR.

I think my stance is "yeah lets discuss this, but we probably don't need to hold up this PR to do it".

**Comment by devil-ira** [comment]
Thanks for the response cart!
Agreed, the current approach based on @JoJoJet's work is nice and a more convenient api can be considered and exist separately.
I'll focus on getting this in a nicer state and I'll try to give @JoJoJet's [PR](https://github.com/bevyengine/bevy/pull/6817) a review :)

**Comment by nicopap** [comment]
Regardless, having the API even if it requires the system param is still precious, since it's easy to build a global variable based abstraction on top of it.

**Comment by djeedai** [comment]
I would expect that:
- a `debug_line!()` API is super convenient for checking a few small stuffs, but is limited in terms of performance and number of items it can draw; this is the equivalent of `println!()` and we know its uses and limitations.
- the current approach is a heavily optimized alternative for heavier uses, like drawing a navmesh or other spatial data structures that have a large number of items and require 10k~100k lines each of more.

I think both use cases are valid and call for different solutions.

File: crates/bevy_gizmos/src/gizmos.rs
Comment (superdump) [inline crates/bevy_gizmos/src/gizmos.rs:80]: What is this for?

File: crates/bevy_gizmos/src/lib.rs
Comment (superdump) [inline crates/bevy_gizmos/src/lib.rs:40]:
```suggestion
const LINE_SHADER_HANDLE: HandleUntyped =
```

File: crates/bevy_gizmos/src/lib.rs
Comment (superdump) [inline crates/bevy_gizmos/src/lib.rs:47]:
```suggestion
        load_internal_asset!(app, LINE_SHADER_HANDLE, "lines.wgsl", Shader::from_wgsl);
```

**Review by superdump — COMMENTED** [review]
Quick first pass.

**Comment by JMS55** [comment]
Taking a quick look, rather than add a render command to the main passes, I think it would be better to have a separate render node that runs after post processing.

File: crates/bevy_gizmos/src/gizmos.rs
Comment (alice-i-cecile) [reply crates/bevy_gizmos/src/gizmos.rs:80]: A comment here would be appreciated.

**Comment by devil-ira** [comment]
> Taking a quick look, rather than add a render command to the main passes, I think it would be better to have a separate render node that runs after post processing.

This is now done, all thanks to @IceSentry :)

**Comment by devil-ira** [comment]
There's currently an issue with UI no longer rendering.

**Review by nicopap — CHANGES_REQUESTED** [review]
Looks good to me, just need to fix the text issue. Plus a single non-blocking comment:

- Should the 2d and 3d APIs be explicitly separated? If so how?

File: crates/bevy_gizmos/src/lib.rs
Comment (devil-ira) [reply crates/bevy_gizmos/src/lib.rs:103]: Sounds good. Added to the 'Future work' section of the PR description.

File: examples/stress_tests/many_gizmos.rs
Comment (devil-ira) [reply examples/stress_tests/many_gizmos.rs:109]: UI isn't rendering. Not sure why. Hopefully @IceSentry can look into it.

**Comment by happydpc** [comment]
How about the line thickness? The lines seem to be too thin.

**Comment by nicopap** [comment]
@happydpc WGPU doesn't give control about line thickness, only 1 pixel. This is very much out of scope for this PR. I think even with only 1px, it's still good enough for debugging. With thicker lines, there is also the issue of line caps sneaking in. I'm not sure what a path toward proper line thickness would look like, most likely using quads, similar to <https://github.com/ForesightMiningSoftwareCorporation/bevy_polyline>.

**Comment by devil-ira** [comment]
> How about the line thickness? The lines seem to be too thin.

Unfortunately line-thickness is not widely supported (OpenGL only I think).
Unity has the same limitation and Unreal instead draws 2 quads in an X shape (like grass in Minecraft) which is pretty janky.

As @nicopap mentioned, line width is going to be tricky and deserves it's own PR.

File: examples/stress_tests/many_gizmos.rs
Comment (devil-ira) [reply examples/stress_tests/many_gizmos.rs:63]: It would be clearer something was going on if the text was rendering :sweat_smile:
I could also make the fancy mode the default (press space to see it)

File: examples/stress_tests/many_gizmos.rs
Comment (IceSentry) [reply examples/stress_tests/many_gizmos.rs:109]: So, the issue is that the node seems to overlap with the ui_pass. I don't fully understand why this happens, but if I schedule it before ui it works correctly. Unfortunately, the ui_pass node is not available in bevy_core_pipeline so we need to hardcode it for now.

**Review by IceSentry — APPROVED** [review]
I'm personally happy with the current state. It's good enough to be useful as is and opens up the potential for future improvements.

**Review by nicopap — APPROVED** [review]
LGTM, works as expected. It's definitively usable in the current state, and it's a greatly useful feature.

**Comment by nicopap** [comment]
bors try

File: crates/bevy_gizmos/src/lib.rs
Comment (cart) [reply crates/bevy_gizmos/src/lib.rs:95]: Yeah I agree that this seems like something that should be per-gizmo. Fine to do in a followup, but in practice I think global `on_top` config will almost never be useful for "real tools", which will need a combination of both.

File: crates/bevy_gizmos/src/pipeline_3d.rs
Comment (cart) [inline crates/bevy_gizmos/src/pipeline_3d.rs:51]: Not something to fix here, but the fact that need to manually plug these values in, despite not using them in the shader, is unfortunate.

**Review by cart — APPROVED** [review]
Fantastic work!

Ultimately I think we'll to explore rendering using methods similar to Freya Holmers [Shapes](https://acegikmo.com/shapes/) lib, as normal "mesh line drawing" has limitations when it comes to thickness, anti-aliasing + quality, etc. Definitely a "hard problem" though. I'm glad we started with the mesh-line approach so we can start building tools + visual debuggers now.

**Comment by cart** [comment]
Turns out this was based on @Toqozz's `bevy_debug_lines` crate, which is licensed under the mit license only. This will need to be re-licensed if we're going to use it in Bevy.

@Toqozz: can you write "I agree to re-license the code from bevy_debug_lines used in this PR to the dual MIT/Apache-2.0 license Bevy uses" here if you agree to the re-license? No pressure btw. It is your code and you can do whatever you want with it. We will revert if we don't hear from you / if the answer is no.

**Comment by Toqozz** [comment]
I agree to re-license the code from bevy_debug_lines used in this PR to the dual MIT/Apache-2.0 license Bevy uses.

**Comment by cart** [comment]
Much appreciated! Thank you for building cool stuff!

**Pull request merged**
